[API] Support dynamic telemetry metadata

larkin opened the pull request:
Add support for dynamic telemetry metadata via custom telemetry metadata providers.  These providers can choose which object types they support, and can implement whatever synchronous logic they would like in order to provide telemetry metadata.  This is intended to better support both type-based metadata and dynamic metadata.

Type-based telemetry metadata is useful for user-created telemetry types like the sinewave generator where metadata can be updated independently from the model properties.

Dynamic telemetry metadata supports usage in summary widgets, where the possible enumerated values depend on the configuration of the summary widget.  It theoretically can support usage in the telemetry filter and the telemetry combiner plugins.

The metadata provider API should be considered unstable in it's current invocation.

Goals of this are:

1. allow summary widgets to emit enumerated telemetry based on configuration, so they can be used as inputs for other summary widgets.
2. to re-enable "type based" telemetry instead of using the type "initializer" (e.g. for SWGs).
3. to simplify telemetry metadata handling for the telemetry filter.

It could be considered a shortfall that this only allows synchronous providers, need some time to think on that.

@akhenry @VWoeltjen Would love a review for discussion purposes.

VWoeltjen commented:
First thought is that separating this out from telemetry providers goes back in the direction of very fine-grained extensions. Moving getMetadata on to TelemetryProvider seems desirable; does this make things messy in cases where you want to use multiple providers for different strategies?

larkin commented:
@VWoeltjen Yes, that's one area where it gets messy.  Since they all have different applicability checks anyhow, it appears logical to separate them.  As we could transition from fine grained to combined and back without impacting backwards compatibility, I don't see us needing to change it at this time.

larkin commented:
Pushed some doc updates.  Thinking that maybe a single addProvider method is okay when you don't have to implement all methods on one provider.  So still supports codesplitting for those that want it.

Going to revise PR to match docs and add limit evaluators too.  Then ready to get this integrated next week.

larkin commented:
@VWoeltjen @akhenry I have revised this based on your feedback and added some additional documentation.  Changes since last review:

* @VWoeltjen after some consideration, you have convinced me to use a single provider instead of multiple APIs.  Optional methods still allow code separation but do not require it.
* added API support for limit evaluators as more optional provider methods.
* plan to depreciate `canProvideTelemetry` and replace with `isTelemetryObject`.  From commit message:
   > Replace `TelemetryAPI.canProvideTelemetry` with
   > `TelemetryAPI.isTelemetryObject`.  This supports general feature
   > checks without requiring a provider to be availalble.  Because
   > application config can change dynamically, there might not always
   > be a request provider or subscription provider, but we can still
   > determine that it is a telemetry object.
   > 
   > The telemetry API will be responsible for providing default
   > functionality when providers are not available for a given request
   > or subscription.
   
   Feedback?

If there are no serious concerns, I'd like to merge this before next testathon.

# Author Checklist

1. Changes address original issue? Y, with some expansion.
2. Unit tests included and/or updated with changes? N, backfill issue still exists.
3. Command line build passes? Y but circle disagrees.
4. Changes have been smoke-tested? Y

VWoeltjen commented on src/api/telemetry/TelemetryAPI.js:
```diff
@@ -1,9 +1,9 @@
 /*****************************************************************************
- * Open MCT, Copyright (c) 2014-2017, United States Government
+ * Open openmct, Copyright (c) 2014-2017, United States Government
```
Find-replace accident here

VWoeltjen commented:
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? N (as noted above)
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

akhenry commented on API.md:
```diff
@@ -1,3 +1,57 @@
+      - [Defining and Registering Time Systems](#defining-and-registering-time-systems)
+      - [Getting and Setting the Active Time System](#getting-and-setting-the-active-time-system)
+    - [Time Bounds](#time-bounds)
+  - [Clocks](#clocks)
```
Should be below Time API. Think it's due to incorrect number of #s in titles

larkin replied:
Fixing.

akhenry commented on API.md:
````diff
@@ -414,7 +478,45 @@ Telemetry requests support time bounded queries. A call to a _Telemetry Provider
+}
+```
+
+This strategy says "I want the lastest data point in this time range".  A provider which recognizes this request should return only one value-- the latest-- in the requested time range.  Depending on your back-end implementation, performing these queries in bulk can be a large performance increase.  These are generally issued by views that are only capable of displaying a single value and only ne ...
````
The size attribute is redundant in this case, we should get rid of it. Unless there is a case for supporting latest N results, in which case we should clarify this in the text.

akhenry commented on API.md:
```diff
@@ -484,6 +586,16 @@ The key-value pairs of this object are described by the telemetry metadata of
+
+### Telemetry Visualization APIs **draft**
+
+The APIs for fetching and visualizing telemetry-- those that would be used in custom views-- are currently in draft state and are being revised.  If you'd like to experiement with them before they are finalized, please contact the  team via the contact-us link on our website.
```
I wonder if perhaps we should provide at least some rudimentary API docs here, with the big caveat that they are draft. We are building views with the new API now, so I think our basic approach is fixed, even though some details might change. It could be as simple as https://github.com/nasa/openmct/issues/1807#issuecomment-344345782

larkin replied:
clarification: andrew refers to "view" api, not telemetry accessors.

larkin replied:
Will clarify this segment to discuss telemetry retrieval api so that it isn't confused with View API.

akhenry commented on example/generator/SinewaveLimitProvider.js:
```diff
@@ -0,0 +1,88 @@
+        YELLOW = 0.5,
+        LIMITS = {
+            rh: {
+                cssClass: "s-limit-upr s-limit-red",
```
Hrm, should our limit evaluators be defining CSS classes? I kind of expect limit evaluators to simply determine whether a given telemetry value is out of limits, and for the view to determine how to visually represent this.

larkin replied:
Concepts such as "red high" and "yellow low" are not universal, as other projects may have "ERROR", "OK" with different display needs.  Additionally, LimitEvaluators can be repurposed as "value stylers" for color coding or other purposes.  Allowing developers to define custom cssClasses allows them to achieve whatever styling they desire.  Plots are an edge case where limitEvaluators influence alarm markers and ignore the specific cssClass; if/when that becomes a problem we will be better equipped to work it out.

akhenry commented on src/api/telemetry/LegacyTelemetryProvider.js:
```diff
@@ -141,17 +141,20 @@ define([
-                evaluate: function () {}
-            };
-        }
+    LegacyTelemetryProvider.prototype.getLimitEvaluator = function (domainObject) {
```
I wonder if there will be a case for a default limit evaluator that reads limits from telemetry metadata? This would make it simple for folks to get up and running with limits, without having to write yet another provider.

larkin replied:
See [comment](https://github.com/nasa/openmct/pull/1941#pullrequestreview-105464455) above, not going to add a default at this time due to this being in flux.

akhenry commented on src/api/telemetry/TelemetryAPI.js:
```diff
@@ -397,8 +446,34 @@ define([
+        var provider = this.findLimitEvaluator(domainObject);
+        if (!provider) {
+            return {
+                evaluate: function () {}
```
See previous comment about a default limit evaluator

larkin replied:
See [comment](https://github.com/nasa/openmct/pull/1941#pullrequestreview-105464455) above, not going to add a default at this time due to this being in flux.

larkin commented:
@akhenry I have updated API docs to address your comments.  Please review and integrate.

larkin commented:
@akhenry now really really ready for review and integrate.
